api/odilo.py

```diff
@@ -0,0 +1,532 @@
+        if modication_date:
+            url = "%s&modificationDate=%s" % (url, modication_date)
+
+        # Retrieve first group of records
+        status_code, headers, content = self.api.get(url)
+        content = json.loads(content)
+
+        # Retrieve Odilo record in groups
+        while status_code == 200 and len(content) > 0:
+            offset += limit
+            retrieved += len(content)
+            self.log.info('Retrieved %i records' % retrieved)
+
+            # Process a bunch of records retrieved
+            for record in content:
+                record_id = record['id']
+                self.log.info('Processing record %i/%i: %s' % (parsed, retrieved, record_id))
+                identifier, is_new = self.api.odilo_bibliographic_coverage_provider.process_item(
+                    record_id, record
+                )
+
+                if is_new:
+                    new += 1
+
+                parsed += 1
+
+            # Persist each bunch of retrieved records
+            self._db.commit()
+
+            # Retrieve next group of records
+            url = "%s?limit=%i&offset=%i" % (self.api.ALL_PRODUCTS_ENDPOINT, limit, offset)
+            if modication_date:
+                url = "%s&modificationDate=%s" % (url, modication_date)
+
+            status_code, headers, content = self.api.get(url)
+            content = json.loads(content)
+
+        if status_code >= 400:
+            self.log.error('ERROR: Fail while retrieving data from remote source: HTTP ' + status_code)
+            if content:
+                self.log.error('ERROR response content: ' + str(content))
+        else:
+            self.log.info('Retrieving all ids finished ok. Retrieved %i records!!' % retrieved)
+            self.log.info('New records: %i' % new)
+
+
+class FullOdiloCollectionMonitor(OdiloCirculationMonitor):
+    """Monitor every single book in the Odilo collection.
+
+    This tells us about books added to the Odilo collection that
+    are not found in our collection.
+    """
+    SERVICE_NAME = "Odilo Full Collection Overview"
+    INTERVAL_SECONDS = 3600 * 4
+
+    def run_once(self, start=None, cutoff=None):
+        """Ignore the dates and return all IDs."""
+        self.log.info("Starting recently_changed_ids, start: " + str(start) + ", cutoff: " + str(cutoff))
+
+        start_time = datetime.datetime.now()
+        self.recently_changed_ids(None)
+        finish_time = datetime.datetime.now()
+
+        time_elapsed = finish_time - start_time
+        self.log.info("recently_changed_ids finished in: " + str(time_elapsed))
+
+
+class RecentOdiloCollectionMonitor(OdiloCirculationMonitor):
+    """Monitor recently changed books in the Odilo collection."""
+
+    SERVICE_NAME = "Odilo Collection Recent Monitor"
+    INTERVAL_SECONDS = 60
+
+
+class MockOdiloAPI(BaseMockOdiloAPI, OdiloAPI):
+    def patron_request(self, patron, pin, *args, **kwargs):
+        pass
```

Comment:
Generally our mock APIs record the fact that a request was made and return a previously queued-up response.